### ISSUE TITLE
Simple Global Secondary Index Query Support

### ISSUE THREAD

**Issue by jrduncans** [d2a4u/meteor#229]
Currently `SecondarySimpleIndex` `retrieve` method returns `F[Option[T]]`. But if you have a simple Global Secondary Index, you can have any partition key (not guaranteed unique value), so `retrieve` should return `Stream[F, T]`. (While a LocalSecondaryIndex would have the same partition key as the table, it must have a sort key, and so wouldn't be used with a `SecondarySimpleIndex`).

Currently having to use a workaround like:

```scala
val byAuthorIndexFixed = SecondaryCompositeIndex[IO, String, String](
    "book_table",
    "book_by_author_GSI",
    KeyDef[String]("author", DynamoDbType.S),
    KeyDef[String]("dummy", DynamoDbType.S),
    dynamoDbAsyncClient
  )
```

**Comment by d2a4u** [comment]
Interesting, I didn't know that there is no uniqueness guarantee on GSI. Changing the signature of `retrieve` to `Stream[F, T]` is a bit weird since `SimpleTable` and local version of `SecondarySimpleIndex` use `F[Option[T]]` and it fits a bit better in those cases. IMHO, the hierarchy should be:
```
sealed abstract class SimpleIndex[F[_]: Async, P: Encoder] {
  ...
  def retrieve[T: Decoder](
    query: Query[P, Nothing],
    consistentRead: Boolean
  )(implicit RT: RaiseThrowable[F]): F[Option[T]]
}

case class SimpleLSI[F[_]: Async, P: Encoder](
  tableName: String,
  indexName: String,
  partitionKeyDef: KeyDef[P],
  jClient: DynamoDbAsyncClient
) extends SimpleIndex[F, P] {
  ...
}

case class SimpleGSI[F[_]: Async, P: Encoder](
  tableName: String,
  indexName: String,
  partitionKeyDef: KeyDef[P],
  jClient: DynamoDbAsyncClient
) extends SimpleIndex[F, P] {
  def retrieveAll[T: Decoder](
    query: Query[P, Nothing],
    consistentRead: Boolean
  )(implicit RT: RaiseThrowable[F]): Stream[F, T]
}
```

**Comment by jrduncans** [comment]
Is a SimpleLSI actually possible? I believe an LSI has to have a sort key.

**Comment by jrduncans** [comment]
I could see either side for only having retrieve return a stream vs have a retrieve/retrieveAll. Allowing a retrieve that returns `F[Option[T]]` that might actually arbitrarily select the last element (and if accidentally really mis-used, only after streaming through many records -- perhaps the implementation should change to be `.head.compile.last`...). But it's also true that someone may have a GSI with a guaranteed-to-be-unique partition key (I have such use-cases myself) and that `retrieve` would read more clearly than having to add `.compile.last`.

**Comment by d2a4u** [comment]
> Is a SimpleLSI actually possible? I believe an LSI has to have a sort key.

that is a good point, I think I've found a fix for this. I moved `SecondarySimpleIndex` to extends `CompositeIndex` in this branch, the signature of `retrieve` fits better there, the only downside is naming. WDYT? https://github.com/d2a4u/meteor/pull/234/files#diff-06c5a4f5c02eebaa484b12addf3b0efc009b69d1e93124501aac37d7dbf31183R59

**Comment by jrduncans** [comment]
Seems ok to me, let's me do what I need:

```scala
byAuthorIndex.retrieve[Book](
        Query(partitionKey = "Jules Verne"),
        limit = 100
      )
```

The hacky SecondaryCompositeIndex in my original post has the nicer ergonomics of not needing to wrap in a query though which would be nice for `SecondarySimpleIndex`:

```scala
byAuthorIndexFixed.retrieve[Book](
        "Jules Verne",  limit = 100
      )
```

**Comment by d2a4u** [comment]
it is a overloading method so the signature has been changed to [this](https://github.com/d2a4u/meteor/pull/234/files#diff-06c5a4f5c02eebaa484b12addf3b0efc009b69d1e93124501aac37d7dbf31183R88)

**Comment by jrduncans** [comment]
Looks good to me!

**Comment by d2a4u** [comment]
This is now fixed in latest version: [v1.0.14](https://github.com/d2a4u/meteor/releases/tag/v1.0.14). It should have been bumped to 1.1.0 but there seems to be an issue with CI versioning bump. Apologies.
